File: content/problems/0722-remove-comments/solutions/block-comment-state-machine/Solution.java
class Solution {
    public List<String> removeComments(String[] source){
        List<String> answer=new ArrayList<>();
        boolean inBlock=false;
        StringBuilder buffer=new StringBuilder();
        for(String line:source){
            for(int index=0;index<line.length();){
                if(inBlock){
                    if(index+1<line.length()&&line.charAt(index)=='*'&&line.charAt(index+1)=='/'){
                        inBlock=false;
                        index+=2;
                    }else index++;
                }else if(index+1<line.length()&&line.charAt(index)=='/'&&line.charAt(index+1)=='/')break;
                else if(index+1<line.length()&&line.charAt(index)=='/'&&line.charAt(index+1)=='*'){
                    inBlock=true;
                    index+=2;
                }else buffer.append(line.charAt(index++));
            }
            if(!inBlock&&buffer.length()>0){
                answer.add(buffer.toString());
                buffer.setLength(0);
            }
        }
        return answer;
    }
}

File: content/problems/0722-remove-comments/solutions/block-comment-state-machine/solution.cpp
class Solution {
    public: vector<string> removeComments(vector<string>& source){
        vector<string> answer;
        bool inBlock=false;
        string buffer;
        for(string& line:source){
            for(int index=0;index<(int)line.size();){
                if(inBlock){
                    if(index+1<(int)line.size()&&line[index]=='*'&&line[index+1]=='/'){
                        inBlock=false;
                        index+=2;
                    }else ++index;
                }else if(index+1<(int)line.size()&&line[index]=='/'&&line[index+1]=='/')break;
                else if(index+1<(int)line.size()&&line[index]=='/'&&line[index+1]=='*'){
                    inBlock=true;
                    index+=2;
                }else buffer+=line[index++];
            }
            if(!inBlock&&!buffer.empty()){
                answer.push_back(buffer);
                buffer.clear();
            }
        }
        return answer;
    }
};
